### ISSUE TITLE
Assignment issue with non-optional mapped type as function argument

### ISSUE THREAD

**Issue by andreytrofimov** [microsoft/TypeScript#32160]
**TypeScript Version:**  3.5.2, 3.6.0-dev.20190627

**Search Terms:** mapped type 2322 strictFunctionTypes

**Code**
(requires --strictFunctionTypes)

```ts
class Type1<T> {
    fn = (arg: { [P in keyof T]-?: any; }): void => void arg;
}

const a: Type1<any> = new Type1<{foo: 'foo'}>();
// Type 'Type1<{ foo: "foo"; }>' is not assignable to type 'Type1<any>'.
//   Types of property 'fn' are incompatible.
//     Type '(arg: { foo: any; }) => void' is not assignable to type '(arg: { [x: string]: any; }) => void'.
//       Types of parameters 'arg' and 'arg' are incompatible.
//         Property 'foo' is missing in type '{ [x: string]: any; }' but required in type '{ foo: any; }'.

a.fn({foo: 'foo'}); // no error
```

But the following cases don't emit errors
```ts
class Type2<T> {
    v!: { [P in keyof T]-?: any; };
}

const b: Type2<any> = new Type2<{foo: 'foo'}>();
```

```ts
class Type3<T> {
    fn = (arg: { [P in keyof T]: any; }) => void arg;
}

const c: Type3<any> = new Type3<{foo: 'foo'}>();
```

```ts
class Type4<T> {
    fn(arg: { [P in keyof T]: any; }) { return void arg; }
}

const d: Type4<any> = new Type4<{foo: 'foo'}>();
```

**Expected behavior:**
`Type1<{foo: 'foo'}>` should be assignable to `Type1<any>`

**Actual behavior:**
`Type1<{foo: 'foo'}>` is not assignable to `Type1<any>`

**Playground Link:** 

(requires --strictFunctionTypes)
[Link](https://typescript-play.js.org/#code/PTAEFpK6dyCgSgMIEMDOBTUBGeBjAGw3VABUBPAB0xwB4yA+UAb3lA9ADMA7UAXlAAKVACcA5gC5WoANoAFUAEs+Aa0wUA9l3IBdcAH5pqHhQDcoAL4BKaQDdNSgCYDmD56DHiz8S-AKaPOgALp7SlDT0JhTMgjyYAO7k1LR0LFyamtIA5Bma2ZaMQtY+SABMAMxlZdKIYBHY2Q30LNyZ0gBEeR0WhdnKpDyaoSRK4jyoAEaE2MGaoMEpoE0pUaaM2QB0dZwNpNqgVKKaNKKLy7z9Ytgq+JoAtlSowUrTmNtInMk0yyIS0q08sZTL1rK5QO4nP0lINhp50OgxhM3gt5osftk-lIZLIAB7SEKiFTiXTA8xWMH8NyOKEfMBfDh7UAHJ6iVD3TDBTCiUjZLxXHguPkSK6iG48O6PZ6vGZ0hlfeTHU7nXKZaGke4wxE8cTKPjoxqtPEE4JEnWkzwgqz9SYAV1CYoAjralGKXCoFktsoD2pbyZYtv5UJteEJ0r7VfkbD5-Eg4PH4zs0FhQGUCMQEd9MGUGMw2F87ABCAFyRQe9RaHRkfRGP29Hx+AJBUKTcIpHPRWKgeJJBo58NZC5qwrFGM7BMT6BJjDYCrpkhZiq51jsTi8ATCLwlhR60AVg7Vsmg8GQzwSBv+O7N0D4Ns0Jedjc9xdpIFDqNFEr+ceTyfTlMACzzpmDQAcu+ZrjwWLbmWagaAeFrRMerRisEtqiHwp5eL0viXoEISgE4d6YGBj5xIkWZgQOOR5AUn4xkAA)

**Related Issues:** 
No related issues found

**Comment by dragomirtitian** [comment]
`Type1` behaves as expected i think. Since you have `strictFunctionTypes` functions behave contravariantly with regard to parameter types. So the assignment the compiler complains about is that `(arg: { foo: any; }) => void`, a function whose parameter requires a property `foo`, is not assignable to a function `(arg: { [x: string]: any; }) => void` that makes no such requirements on the presence of `foo`. 

Passing `any` though the mapped type will make it lose its special status of being assignable to and from anything and will just transform it to `{ [x: string]: any; }` in this case.

`Type2` does not produce an error because `T` is used in a covaraint position (a field)
`Type4` does not produce an error because in spite of `strictFunctionTypes` methods still behave bivariantly with regard to parameter types

The only surprise I have is that `Type3` does not produce an error. Not adding the `-?` seems to change the result of the mapping `any` in some non obvious way.

**Workaround**
If `T` is in a contravariant position, using `never` instead of `any` will allow the assignment, since mapping `never` produces `never` in this case (and anything is assignable to `never`)

```ts
    class Type1<T> {    
        fn = (arg: { [P in keyof T]-?: any; }): undefined => void arg;
    }

    const a: Type1<never> = new Type1<{foo: 'foo'}>();
```

**Comment by andreytrofimov** [comment]
@dragomirtitian 
Thank you for your explanation!

I have more complicated case, therefore, I only have an **invariant** way

```ts
interface Validator<T> { (v: T): false | string }

interface Field<T> {
    value: T;
} 

interface Form<T> {
    value: T;
    validator: Validator<T>;
}

type FormFromType<T> = Form<{ [P in keyof T]-?: Field<T[P]> }>;

declare let form: any

type FooForm = FormFromType<{foo: Field<'foo'>}>;
type AnyForm = FormFromType<any>;
type NeverForm = FormFromType<never>;

const c1: AnyForm = form as FooForm; // error
const c2: NeverForm = form as FooForm; // error
const c3: FooForm = form as AnyForm; // error
const c4: FooForm = form as FooForm; // ok

```
